075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,500 @@
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance-template
+  annotations:
+    strimzi.io/rebalance: template # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  # NOTE: mode and brokers fields, if set, will be just ignored because they are
+  #       automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+When the `spec.cruiseControl.autoRebalance` is set and a specific rebalancing mode refers to a "template" `KafkaRebalance` custom resource, the operator checks that it exists in the first place.
+If it doesn't, the operator logs a warning about auto-rebalancing being ignored (as it wasn't set at all) because referring to a non existing "template" `KafkaRebalance` custom resource.
+In this case, the scaling process proceeds as usual: the scale up just runs (without any following rebalance) while the scale down could be blocked (because no rebalance ran to get topics' partitions off the brokers to remove).
+
+If the referenced "template" `KafkaRebalance` custom resource exists, the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource when the user scales the cluster, by adding and/or removing brokers.
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` and `spec.brokers` to it, based on the scaling operation the user asked for and on which brokers.
+This is the reason why the same fields specified in the "template" `KafkaRebalance` custom resource are ignored.
+The "actual" `KafkaRebalance` custom resource will be named as `<my-cluster-name>-auto-rebalancing-<mode>` where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and the `<mode>` is referring to which rebalancing has to run (add or remove brokers).
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-cluster-auto-rebalancing-add-brokers
+  finalizers:
+    - strimzi.io/auto-rebalancing
+spec:
+  mode: add-brokers
+  brokers: [3,4]
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+
+The operator also sets a finalizer, named `strimzi.io/auto-rebalancing`, on the "actual" `KafkaRebalance` custom resource.
+This is needed to avoid the user, or any other tooling, to delete the resource while the auto-rebalancing is still running.
```

Comment:
What if the actual Rebalance is deleted despite the finalizer? This can be done when the operator is not running. When the operator restarts, I guess we need to reset the auto rebalance to idle state and try again, as you don't know the outcome of that Rebalance.

Reply:
> What if the actual Rebalance is deleted despite the finalizer? This can be done when the operator is not running

Not sure Kubernetes works this way but I could be wrong. If you have a finalizer, it doesn't matter if the component which set the finalizer is running or not (the operator in this case), you cannot delete the custom resource. I assume you can force removing the finalizer and then deleting the resource, but you are doing something like this on purpose, so you deserve the process not working anymore imho. We can't fight with all corner cases made by (stupid) users.

Said that what I can do is that on the reconciliation, if when checking the `KafkaRebalance` status for an ongoing rebalancing, the operator detects that it's missing, the autorebalance state machine moves to `NotReady`. From there, it's all in the user hands as already discussed with @scholzj which seems to be the best solution given there is nothing more we can do.

Reply:
> It's the operator that sets the finalizer, and if it's not running nobody re-adds the finalizer and the resource can be deleted.

From your initial comment you said, the operator is not running anymore and someone delete the `KafkaRebalance`. If the operator is stopped/crashed, no one remove the finalizer. So when user tries to delete the resource, it's not possible to do so because the finilizer is still there. Of course, there should be a way to force deletion but it would be you to put yourself in trouble.
So I don't get what your comment about "and if it's not running nobody re-adds the finalizer" ... the finalizer is still there if operator crashes/stopped.

Reply:
I was talking about the evil user, who first stops the operator, and then removes the finalizer by simply editing the resource and deleting that row. I don't know if that's what you mean with "force".

Reply:
Yeah that was my assumption. If you are deleting a finalizer, for deleting the resource, you should be aware to put yourself in troubles tbh. The finalizer is there for a reason ... to avoid the resource being deleted when it should not be and you are forcing that limit. So I don't expect we helping you.